---
title: "Diallel cross: A case study of mating designs"
author: "Deependra Dhakal"
date: '2018-09-15'
# categories: "Plant Breeding"
header:
  caption: ''
  image: ''
link-citations: yes
math: yes
slug: diallel-cross-analysis
tags:
- agriculture
- research-review
bibliography: ../bib/2018-09-15-diallel-cross-analysis.bib
---


```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(cache = TRUE, echo = FALSE, 
                      tidy.opts = list(width.cutoff=50), 
                      eval = TRUE, fig.align = "center", fig.asp = 0.6, 
                      message = FALSE, warning = FALSE, cache = TRUE)
options(knitr.kable.NA = "", digits = 3, knitr.table.format = "html")
library(tidyverse)
require(plantbreeding)
set.seed(453)
```

# Mating designs

## Introduction

Mating designs allow for partitioning of phenotypic effects -- as due to genotype, environment or interacting effects among genes and alleles. Using one or more of these mating schemes, identification of heterotic groups, estimation of general and specific combining abilities and testing of environmental interactions could be done. Progenies resulting from a well designed mating are used for the dissection of trait genetics.

In order to understand genetics of traits and to make effective choice of parents, two contrasting methods of selection should be understood first. Bluntly, selection methods can be stated as either being forward or backward. Forward selection is synonymous to **within-family selection** whereas the concept of backward selection embodies **selection among families**. Ideally, forward selection works best for highly heritable traits -- for those traits regulated by few small genes, as opposed to those involving large number of genes with small cumulative effects. 

Selection gets more complicated as data on several different individuals belonging to some families is available. [What we study versus what reality is; Insert soil layer image here]. Broadly, three distinct modes of selection could be practiced:

1. Strict within family selection
2. Selection on within family deviation
3. Combined family and within family selection; family selection index

When the interest is to exploit the state of heterosis arising from certain combination of parental individuals, the genetic factors contributing well to superior phenotype should be underpinned. The whole process of determining favorable combination among parental individuals should be met with phenotypic data from many progeny, which is retrospective in purpose -- thus the name backward selection. Family selection has different variants and serve variying purposes as well. 

- Half-sib selection is used to select superier individuals for their GCA.
- Full-sib selection is used to make distinct parental matings in order to induce hybrid vigor by capturing specific combinining abilities

The concept of combining abilities was first laid out by Sprague and Tatum in 1942 [@sprague1942general] in order to generate variance estimates without too much of underlying genetic assumptions. The combining ability test procedure involves making crossess of several different combinations from a set of parents and ascribing the resultant variances statistically to either the genetic additiveness of parental charactersistics or the interacting parental genetic combinations. Thus, the phenotype ($y_i$) of a cross progeny can be modeled as linear combination of additive ($A_i$), dominance ($D_i$) and environmental ($e_i$) effects, is:

$$y_i = \mu + A_i + D_i + e_i$$

## Combining ability

Crossing each line with several other lines produces an additional measure in the mean performance of each line in all crosses. This mean performance of a line, when expressed as a deviation from the mean of all crosses, gives what is called the general combining ability (**GCA**) of the lines. The GCA is calculated as the average of all $F_1$s having this particular line as one parent, the value being expressed as a deviation from the overall mean of crosses. Each cross has an expected value (the sum of GCAs of its two parental lines).

The mean genotypic value of offspring from a particular cross may deviate from value expected considering the population mean and the sum of the parental GCA effects. This deviation is the specific combining ability (**SCA**) for that cross.

The differences of GCA are due to the additive and additive x additive interactions in the base population. The differences in SCA are attributable to nonadditive genetic variance. SCA effects are derived from inter-allelic/intra-loci interactions. Furthermore, the SCA is expected to increase in variance more rapidly as inbreeding in the population reaches high levels. GCA is the average performance of a plant in a cross with different tester lines, while SCA measures the performance of a plant in a specific combination in comparison with other cross combinations.

We can define the mean genotypic value ($G_{AB}$) for the full-sib family produced by crossing parents A and B as the sum of the overall mean $\mu$, the GCAs of the two parents and the SCA value:

$$G_{AB} = \mu + GCA_A + GCA_B + SCA_{AB}$$

The types of interactions that can be obtained (SCA effects) depend upon the mating scheme used to produce the crosses, the most common being the diallel mating design, developed by B. Griffing (1956). Methods such as top cross and poly-cross are also not uncommon. A classical method to estimate dominance genetic variance (D) is to estimate the variance associated with SCA effects of many crosses. The expected value of the observed SCA variance component is 1/4 of the dominance genetic variance in the reference population.

The GCA of each line is calculated as follows:

$$
\mathrm{G_x} = \left[\frac{T_x}{n-2}\right]-\left[\frac{\sum T}{n(n-2)}\right]
$$

Where $x$ represents a specific line. Using fabricated dataset given in Table \@ref(tab:fabricated-diallel) following procedures outlines how GCA for Parent 2 (P2) ($GCA_{P2}$) can be calculated.

\begin{equation}
\begin{split}
\mathrm{G_b} & = \left[\frac{T_x}{n-2}\right]-\left[\frac{\sum T}{n(n-2)}\right] \\
 & = \left[\frac{39.7}{8}\right]-\left[\frac{324.7}{10\times 8}\right] \\
 & = 4.96-4.06 \\
 & = 0.9
\end{split}
(\#eq:gca-calculation)
\end{equation}

```{r fabricated-diallel}
# Fabricated half diallel mating data

ten_ind <- gl(10, k = 1, ordered = T, labels = paste("P", 1:10, sep = ""))
mating_mat <- cross_df(.l = list(M = ten_ind, F = ten_ind)) %>% 
  unite(col = "cross", sep = " x ", remove = F) %>% 
  spread(key = F, value = cross) %>% 
  dplyr::rename(Parents = M)

mating_mat[,-1][upper.tri(mating_mat[,-1], diag = TRUE)] = NA

# generate a vector of population means of length = n, representing each parental line
mu_list = rnorm(10, mean = 3.5, sd = 1.0)
  
# generate a vector of sds of length = n, representing each parental line
sd_list = rnorm(10, mean = 1.2, sd = 0.3)

# generate a dataframe of random numbers for each column
mating_mat_num <- map2_dfc(.x = mu_list, 
     .y = sd_list, 
     .f = ~cbind(rnorm(10, .x, .y))) %>% 
  magrittr::set_colnames(paste("P", 1:10, sep = ""))
mating_mat_num <- replace(mating_mat_num, !map_df(mating_mat[,-1], ~!is.na(.x)), NA)

mating_mat_num %>%
  add_column(.before = 1, Parents = mating_mat$Parents) %>% 
  knitr::kable(caption = "Fabricated data from a diallel cross scheme using 10 parents") %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", font_size = 10) %>% 
  kableExtra::column_spec(1, bold = TRUE) %>% 
  kableExtra::column_spec(column = 1:11, width = "7em")
```

Taking the above table of diallel cross data, total of each individual parental line could be computed by summing over all the crossess involving the common parent. Similarly, the grand totals could be 
obtained by adding together all the individual parents' total. The individual parents' sum and grand total is shown in the Table \@ref(tab:sum-over-ind) below.

```{r sum-over-ind}
as.matrix(mating_mat_num) %>% 
  Matrix::forceSymmetric("L") %>% 
  as.matrix() %>% 
  colSums(na.rm = TRUE) %>% 
  as_data_frame() %>% 
  add_column(.before = 1, Parents = mating_mat$Parents) %>% 
  dplyr::rename("Line total" = "value") %>% 
  add_case(`Parents` = "Total", `Line total` = sum(.$`Line total`)) %>% 
  knitr::kable(caption = "Totals of individual lines and grand total of diallel cross\n scheme using 10 parents") %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", font_size = 12) %>% 
  kableExtra::column_spec(1, bold = TRUE)
```

## Types of diallel mating design

There are four types of popular diallel crossing designs:

1. Full diallels: All the possible combinations of crosses among parents, including reciprocals and self-fertilization of the parents are made. For a sample of $n$ parents, the full-diallel requires $n \times n$ progenies, a number that quickly becomes unmanageable as more parents are sampled (Table \@ref(tab:full-diallel)).

```{r full-diallel}
# Full diallel mating scheme

ten_ind <- gl(10, k = 1, ordered = T, labels = paste("P", 1:10, sep = ""))
mating_mat <- cross_df(.l = list(M = ten_ind, F = ten_ind)) %>% 
  unite(col = "cross", sep = " x ", remove = F) %>% 
  spread(key = F, value = cross) %>% 
  dplyr::rename(Parents = M)

mating_mat %>%
  knitr::kable(caption = "Full diallel mating scheme using 10 parents") %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", font_size = 10) %>% 
  kableExtra::column_spec(1, bold = TRUE) %>% 
  kableExtra::column_spec(column = 1:11, width = "7em")
```


2. Half diallels: Each parent is mated with every other parent, excluding selfs and reciprocals. This requires making $\frac{n(n-1)}{2}$ crosses for n parents (Table \@ref(tab:half-diallel)).

```{r half-diallel}
# Half diallel mating scheme

ten_ind <- gl(10, k = 1, ordered = T, labels = paste("P", 1:10, sep = ""))
mating_mat <- cross_df(.l = list(M = ten_ind, F = ten_ind)) %>% 
  unite(col = "cross", sep = " x ", remove = F) %>% 
  spread(key = F, value = cross) %>% 
  dplyr::rename(Parents = M)

mating_mat[,-1][upper.tri(mating_mat[,-1], diag = TRUE)] = NA
  
mating_mat %>%
  knitr::kable(caption = "Half diallel mating scheme using 10 parents") %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", font_size = 10) %>% 
  kableExtra::column_spec(1, bold = TRUE) %>% 
  kableExtra::column_spec(column = 1:11, width = "7em")
```


3. Partial diallel: Not all the crosses are made. There are no reciprocals or selfs. The goal is to reduce the breeding workload for a given sample of parents by making less than $\frac{n(n-1)}{2}$ crosses for n parents (Table \@ref(tab:partial-diallel)).

```{r partial-diallel}
# Partial diallel mating scheme

ten_ind <- gl(10, k = 1, ordered = T, labels = paste("P", 1:10, sep = ""))
mating_mat <- cross_df(.l = list(M = ten_ind, F = ten_ind)) %>% 
  unite(col = "cross", sep = " x ", remove = F) %>% 
  spread(key = F, value = cross) %>% 
  dplyr::rename(Parents = M)

mating_mat[,-1][upper.tri(mating_mat[,-1], diag = TRUE)] = NA
mating_mat[,-1] <- mating_mat[,-1] %>% 
  mutate_all(.funs = funs(replace(., sample(1:10, 
                                            size = sample(1:5, size = 1, replace = TRUE), 
                                            replace = FALSE), NA)))

# mating_mat[sample(1:10, 2, replace = TRUE), sample(1:10, 3, replace = TRUE)]
  
mating_mat %>%
  knitr::kable(caption = "Partial diallel mating scheme using 10 parents") %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", font_size = 10) %>% 
  kableExtra::column_spec(1, bold = TRUE) %>% 
  kableExtra::column_spec(column = 1:11, width = "7em")
```


4. Connected diallels: Two groups (for example, $1-6$ and $7-12$) of individuals are used to form two diallels (generally partial or other diallel scheme in each group) but they are connected by crossing $4 \times 9$, $7 \times 1$, $9 \times 3$ and $10 \times 2$. In the example below, the second diallel also includes some selfs $(S)$ and reciprocals $(R)$ (Table \@ref(tab:connected-diallel)).

```{r connected-diallel}
# Connected diallel mating scheme

twelve_ind <- gl(12, k = 1, ordered = T, labels = paste("P", 1:12, sep = ""))

# generate independent diallel from first six parents
mating_mat1 <- cross_df(.l = list(M = twelve_ind[1:6], F = twelve_ind[1:6])) %>% 
  unite(col = "cross", sep = " x ", remove = F) %>% 
  spread(key = F, value = cross) %>% 
  dplyr::rename(Parents = M)

# create partial diallel from first matrix
mating_mat1[,-1][upper.tri(mating_mat1[,-1], diag = TRUE)] <- NA
mating_mat1[,-1] <- mating_mat1[,-1] %>% 
  mutate_all(.funs = funs(replace(., sample(1:length(which(!is.na(.))), 
                                            size = sample(1:4, size = 1, replace = TRUE), 
                                            replace = TRUE), NA)))

# generate independent diallel from next six parents
mating_mat2 <- cross_df(.l = list(M = twelve_ind[7:12], F = twelve_ind[7:12])) %>% 
  unite(col = "cross", sep = " x ", remove = F) %>% 
  spread(key = F, value = cross) %>% 
  dplyr::rename(Parents = M)

# create random diallel crosses from second matrix
mating_mat2[,-1] <- mating_mat2[,-1] %>% 
  mutate_all(.funs = funs(replace(., sample(1:6, 
                                            size = sample(3:5, size = 1, replace = TRUE), # because at least 3 diallel crosses will be have to be avoided; likewise minimum of (6-5=1) crosses must be performed for some diallel parent.
                                            replace = FALSE), NA)))

# combine two matrices diagonally using Matrix package
mating_mat <- full_join(mating_mat1, mating_mat2)

# add connecting crossess between two diallel sets
mating_mat[,-1][,c(9, 1, 3, 2)] <- pmap_df(list(mating_mat[,-1][,c(9, 1, 3, 2)], 
             c(4, 7, 9, 10), 
             c("P4 x P9", "P7 x P1", "P9 x P3", "P10 x P2")),
        .f = function(x, y, z)replace(x, y, z))

# see the table  
mating_mat %>%
  knitr::kable(caption = "Connected diallel mating scheme using 12 parents") %>% 
  kableExtra::kable_styling(bootstrap_options = "striped", font_size = 10) %>% 
  kableExtra::column_spec(1, bold = TRUE) %>% 
  kableExtra::column_spec(column = 1:11, width = "7em")
```


With the convenience of using individuals as both male and female parents, diallel mating designs are popular for plant breeding studies. Certain diallel designs allow for estimation of reciprocal cross effects. Diallels cannot be used in dioecious species (female and male flowers occur in different plants). However, factorial designs can be used in dioecious species to estimate dominance genetic variance.

If there are no connections between groups of parents, the design is a diallel in sets. Diallel mating designs provide good evaluation of parents and full-sib families. They also provide estimates of both additive and dominance genetic effects, and genetic gains due to additive and dominance genetic effects if we assume the sample of parents used is sufficient to represent the reference population [@baker1978issues; @holland2003estimating]. One disadvantage of diallels is that the breeding and progeny evaluations can be costly due to large number of crosses required. As evident, For a full diallel with 6 parents, 36 crosses are required; with 12 parents the number of crosses required is `r 12^2`. On the other hand there is the looming question, if the sample of 6 or 12 individuals from a population provides useful estimate of the reference population genetic variances [@baker1978issues]. @white2007forest and @hallauer1988quantitative have described several other forms of diallel mating designs in detail.


<!-- ## Use of sparse matrices in mating designs -->

```{r sparse-matrices, echo=FALSE, include=FALSE}
# One may first learn the concept and use of sparse matrix
flev <- c(1,3:8)
Matrix::sparseMatrix(seq_along(flev), flev, x = 1)

```

<!-- ## Analysis of diallel cross -->

```{r diallel-analysis, echo=FALSE, eval=FALSE, include=FALSE}
## use plantbreeding package
resd <- diallele1(data(fulldial, package = "plantbreeding"), yvar = "YIELD", 
                  progeny = "TRT", male = "MALE", female = "FEMALE", 
                  replication = "REP")

resd$gca.effmat
resd$sca.effmat

## manually summarize the results of diallel mating scheme
map_dbl(fulldial[,-7], ~length(unique(.x)))
fulldial %>% 
  group_by(MALE, FEMALE) %>% 
  summarise_at("YIELD", funs(mean, .args = list(na.rm = TRUE))) %>% 
  spread(key = FEMALE, value = YIELD)
```


## References
